### ISSUE TITLE
inconsistent module to module message throughput

### ISSUE THREAD

**Issue by peirens-bart** [Azure/iotedge#2915]
## Expected Behavior

When module A is sending 1000messages/second to module B and module B is forwarding that message to module C I expect the same time delay between the modules.

## Current Behavior
the delay between module A and B is growing over time while the delay between module B and module C is more stabel

## Steps to Reproduce
Provide a detailed set of steps to reproduce the bug.
1.Create a module A that is sending 1000m/s from multiple threads, this module will log the datetime when sending
2.Create a module B that can forward these messages, the module will log the dattime when receiving and when sending
3.Create a module C that can receive these messages, thi smodule will log the datatime when receiving
4. start sending messages from module A
5. the diff in milliseconds from B-receive - A-Send will grow, while teh diff in milliseconds between C-recieve and B-send stays the same.

## Context (Environment)
my context is that module a is a RFID reader module, module B is a message queue module and module C is a json text builder. module B will receive other events about gps and we need to merge those events with the rfid events so it can be used by an ML module later on.
teh problem is the delay. we need to make sure that the correct gps events are added with the rfid read based on the time.

### Output of `iotedge check`

Configuration checks
--------------------
√ config.yaml is well-formed - OK
√ config.yaml has well-formed connection string - OK
√ container engine is installed and functional - OK
√ config.yaml has correct hostname - OK
√ config.yaml has correct URIs for daemon mgmt endpoint - OK
√ latest security daemon - OK
√ host time is close to real time - OK
√ container time is close to host time - OK
‼ DNS server - Warning
    Container engine is not configured with DNS server setting, which may impact connectivity to IoT Hub.
    Please see https://aka.ms/iotedge-prod-checklist-dns for best practices.
    You can ignore this warning if you are setting DNS server per module in the Edge deployment.
‼ production readiness: certificates - Warning
    The Edge device is using self-signed automatically-generated development certificates.
    They will expire in 84 days (at 2020-07-28 10:05:10 UTC) causing module-to-module and downstream device communication to fail on an active deployment.
    After the certs have expired, restarting the IoT Edge daemon will trigger it to generate new development certs.
    Please consider using production certificates instead. See https://aka.ms/iotedge-prod-checklist-certs for best practices.
√ production readiness: container engine - OK
‼ production readiness: logs policy - Warning
    Container engine is not configured to rotate module logs which may cause it run out of disk space.
    Please see https://aka.ms/iotedge-prod-checklist-logs for best practices.
    You can ignore this warning if you are setting log policy per module in the Edge deployment.
‼ production readiness: Edge Agent's storage directory is persisted on the host filesystem - Warning
    The edgeAgent module is not configured to persist its /tmp/edgeAgent directory on the host filesystem.
    Data might be lost if the module is deleted or updated.
    Please see https://aka.ms/iotedge-storage-host for best practices.
√ production readiness: Edge Hub's storage directory is persisted on the host filesystem - OK

Connectivity checks
-------------------
√ host can connect to and perform TLS handshake with IoT Hub AMQP port - OK
√ host can connect to and perform TLS handshake with IoT Hub HTTPS / WebSockets port - OK
√ host can connect to and perform TLS handshake with IoT Hub MQTT port - OK
√ container on the default network can connect to IoT Hub AMQP port - OK
√ container on the default network can connect to IoT Hub HTTPS / WebSockets port - OK
√ container on the default network can connect to IoT Hub MQTT port - OK
√ container on the IoT Edge module network can connect to IoT Hub AMQP port - OK
√ container on the IoT Edge module network can connect to IoT Hub HTTPS / WebSockets port - OK
√ container on the IoT Edge module network can connect to IoT Hub MQTT port - OK

19 check(s) succeeded.
4 check(s) raised warnings. Re-run with --verbose for more details.
<details>
<summary>Click here</summary>

```

<Paste here between the triple backticks>

```
</details>

### Device Information
* Host OS Ubuntu 18.04
* Architecture amd64,
* Container OS Linux containers

### Runtime Versions
mobile@InWareDrones-NUC10i7FNH:~$ iotedge list
NAME             STATUS           DESCRIPTION      CONFIG
buildtagdata     running          Up 6 hours       hertz.azurecr.io/edgemodule/stringbuilder-amd64:1.0.02
edgeAgent        running          Up 9 hours       mcr.microsoft.com/azureiotedge-agent:1.0
edgeHub          running          Up 13 minutes    mcr.microsoft.com/azureiotedge-hub:1.0
flightcache      running          Up 3 hours       hertz.azurecr.io/edgemodule/eventcache-amd64:1.0.04
mqtoarta         running          Up 6 hours       hertz.azurecr.io/edgemodule/messagequeue-amd64:1.0.03
tmreader         running          Up 3 hours       hertz.azurecr.io/edgemodule/thingmagic-amd64:1.0.03

Note: when using Windows containers on Windows, run `docker -H npipe:////./pipe/iotedge_moby_engine version` instead

## Logs
merge of our logs and edge hub logs

2020-05-05 12:30:10.232 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received method invoke call from iwd-ubuntu-0001/tmreader for iwd-ubuntu-0001/tmreader with correlation ID 0bcdcc49-8c1f-4ff9-94d5-ae9a7d4f3b5f

2020-05-05 12:30:10.2401 | I |  80 | HertzReceiveActor.cs      | 137  | Receive message with Id 0 and Type StartStopReaderEvent from Module DirectMethod in Module tmreader with arguments [{"START": "True"},{"DURATION": "5000"}] 

2020-05-05 12:30:10.2712 | I |  80 | HertzReceiveActor.cs      | 214  | Sending message with Id 6663414471546904576 and Type RfidReadEvent from Module tmreader with arguments [{"ANTENNAPORTNUMBER": "8"},{"RSSI": "-47"},{"TAG": "

2020-05-05 12:30:10.287 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received 1 message(s) from iwd-ubuntu-0001/tmreader

2020-05-05 12:30:10.324 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Storage.MessageStore] - Added message 4f756815-d946-43c3-85f2-5ad08fe81acf to store for iwd-ubuntu-0001/flightcache/DequeueEvent at offset 0 - messageCount = 0

2020-05-05 12:30:10.329 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Storage.MessageStore] - Getting next batch for endpoint iwd-ubuntu-0001/flightcache/DequeueEvent starting from 0 with batch size 10.

2020-05-05 12:30:10.351 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Storage.MessageStore] - Obtained next batch for endpoint iwd-ubuntu-0001/flightcache/DequeueEvent with batch size 2. Next start offset = 2.

2020-05-05 12:30:10.351 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Storage.MessageStore] - Getting next batch for endpoint iwd-ubuntu-0001/flightcache/DequeueEvent starting from 2 with batch size 8.

2020-05-05 12:30:10.351 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Storage.MessageStore] - Obtained next batch for endpoint iwd-ubuntu-0001/flightcache/DequeueEvent with batch size 6. Next start offset = 8.
2020-05-05 12:30:10.351 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Storage.MessageStore] - Getting next batch for endpoint iwd-ubuntu-0001/flightcache/DequeueEvent starting from 8 with batch size 2.
2020-05-05 12:30:10.351 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Storage.MessageStore] - Obtained next batch for endpoint iwd-ubuntu-0001/flightcache/DequeueEvent with batch size 2. Next start offset = 10.
2020-05-05 12:30:10.379 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.ModuleEndpoint] - Sending 10 message(s) to module iwd-ubuntu-0001/flightcache.
2020-05-05 12:30:10.390 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Device.DeviceMessageHandler] - Sent message with correlation ID 26bd2187-519e-4439-a687-87dfb953775c to iwd-ubuntu-0001/flightcache
2020-05-05 12:30:10.392 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Amqp.ClientConnectionHandler] - Sending telemetry message to iwd-ubuntu-0001/flightcache

2020-05-05 12:30:10.3974 | I |  44 | HertzReceiveActor.cs      | 137  | Receive message with Id 6663414471546904576 and Type DequeueEvent from Module tmreader in Module flightcache with arguments [{"ANTENNAPORTNUMBER": "8"},{"RSSI": "-47"},{"TAG": "E2004102691201851680680F"},{"TAGLASTSEEN": "2020-05-05T12:30:10.2710555Z"},{"PHASE": "90"},{"FREQUENCY": "917500"},{"READERTEMPERATURE": "27"},{"READERNAME": "tmreader"},{"READERTIME": "19"},{"READERREADCOUNT": "1"},{"READERPERFORMANCE": "53"},{"ID": "6663414471546904576"}] 

2020-05-05 12:30:10.3974 | I |  44 | HertzReceiveActor.cs      | 214  | Sending message with Id 6663414471546904576 and Type DequeueEvent from Module flightcache with arguments [{"ANTENNAPORTNUMBER": "8"},{"RSSI": "-47"},{"TAG": "E2004102691201851680680F"},{"TAGLASTSEEN": "2020-05-05T12:30:10.2710555Z"},{"PHASE": "90"},{"FREQUENCY": "917500"},{"READERTEMPERATURE": "27"},{"READERNAME": "tmreader"},{"READERTIME": "19"},{"READERREADCOUNT": "1"},{"READERPERFORMANCE": "53"},{"ID": "6663414471546904576"},{"AMOUNTOFNEXTVALUES": "-1"}] 

2020-05-05 12:30:10.397 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received 1 message(s) from iwd-ubuntu-0001/flightcache

2020-05-05 12:30:10.397 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Storage.MessageStore] - Added message bff00c13-5e74-43b8-950b-7ba508d31e79 to store for iwd-ubuntu-0001/buildtagdata/StringBuilderEvent at offset 0 - messageCount = 0
2020-05-05 12:30:10.397 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Amqp.LinkHandlers.EventsLinkHandler] - EventsLinkHandler processed 1 messages for iwd-ubuntu-0001/flightcache
2020-05-05 12:30:10.397 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Storage.MessageStore] - Getting next batch for endpoint iwd-ubuntu-0001/buildtagdata/StringBuilderEvent starting from 0 with batch size 10.
2020-05-05 12:30:10.397 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Storage.MessageStore] - Obtained next batch for endpoint iwd-ubuntu-0001/buildtagdata/StringBuilderEvent with batch size 1. Next start offset = 1.
2020-05-05 12:30:10.398 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Storage.MessageStore] - Getting next batch for endpoint iwd-ubuntu-0001/buildtagdata/StringBuilderEvent starting from 1 with batch size 9.
2020-05-05 12:30:10.398 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Storage.MessageStore] - Obtained next batch for endpoint iwd-ubuntu-0001/buildtagdata/StringBuilderEvent with batch size 0. Next start offset = 1.
2020-05-05 12:30:10.398 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.ModuleEndpoint] - Sending 1 message(s) to module iwd-ubuntu-0001/buildtagdata.
2020-05-05 12:30:10.398 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Device.DeviceMessageHandler] - Sent message with correlation ID b6b3a7ce-c14d-4a1b-9036-45456fe7367b to iwd-ubuntu-0001/buildtagdata
2020-05-05 12:30:10.398 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Amqp.ClientConnectionHandler] - Sending telemetry message to iwd-ubuntu-0001/buildtagdata
2020-05-05 12:30:10.398 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Amqp.LinkHandlers.SendingLinkHandler] - Sent message with id b6b3a7ce-c14d-4a1b-9036-45456fe7367b to device iwd-ubuntu-0001/buildtagdata

2020-05-05 12:30:10.3986 | I |  66 | HertzReceiveActor.cs      | 137  | Receive message with Id 6663414471546904576 and Type StringBuilderEvent from Module flightcache in Module buildtagdata with arguments [{"ANTENNAPORTNUMBER": "8"},{"RSSI": "-47"},{"TAG": "E2004102691201851680680F"},{"TAGLASTSEEN": "05/05/2020 12:30:10"},{"PHASE": "90"},{"FREQUENCY": "917500"},{"READERTEMPERATURE": "27"},{"READERNAME": "tmreader"},{"READERTIME": "19"},{"READERREADCOUNT": "1"},{"READERPERFORMANCE": "53"},{"ID": "6663414471546904576"},{"AMOUNTOFNEXTVALUES": "-1"}] 
2020-05-05 12:30:10.3986 | I |  66 | HertzReceiveActor.cs      | 214  | Sending message with Id 6663414471546904576 and Type StringBuilderEvent from Module buildtagdata with arguments [{"ANTENNAPORTNUMBER": "8"},{"RSSI": "-47"},{"TAG": "E2004102691201851680680F"},{"TAGLASTSEEN": "05/05/2020 12:30:10"},{"PHASE": "90"},{"FREQUENCY": "917500"},{"READERTEMPERATURE": "27"},{"READERNAME": "tmreader"},{"READERTIME": "19"},{"READERREADCOUNT": "1"},{"READERPERFORMANCE": "53"},{"ID": "6663414471546904576"},{"AMOUNTOFNEXTVALUES": "-1"},{"RESPONSE": "{"tagData":{"tag":"E2004102691201851680680F","rssi":"-47","tagLastSeen":"2020-05-05T12:30:10.2710555+00:00","phase":"90","frequency":"917500"},"flightData":$CACHECONTENT;,"readerInfo":{"temperature":"..."}] 

<details>
<summary>iotedged logs</summary>

```

<Paste here between the triple backticks>

```
</details>

<details>
<summary>edge-agent logs</summary>

```

<Paste here between the triple backticks>

```
</details>

<details>
<summary>edge-hub logs</summary>

```

<Paste here between the triple backticks>

```
</details>

## Additional Information

I created a csv (result.text) file with my timestamp

Index: index number
SF: internal snowflake to follow the event trace
E1: module A sending event to module B
E2: module B receiving event from module A
E3: module B sending event to module C
E4: module C receiving event from module B
E5: module C sending event to module D

E2 - E1 is growing
E4 - E3 is more stabel

![image](https://user-images.githubusercontent.com/18715894/80985901-19c2e980-8e30-11ea-9f30-72a6bd277848.png)

is there any reason why this is happening, and can I do anything about it by changing the hub configuration.

```
"edgeHub": {
                        "settings": {
                            "image": "mcr.microsoft.com/azureiotedge-hub:1.0",
                            "createOptions": "{\"User\":\"root\",\"HostConfig\":{\"LogConfig\":{\"Type\":\"json-file\",\"Config\":{\"max-size\":\"10m\",\"max-file\":\"3\"}},\"Binds\":[\"/etc/iotedge/storage/:/iotedge/storage/\"],\"PortBindings\":{\"443/tcp\":[{\"HostPort\":\"443\"}],\"5671/tcp\":[{\"HostPort\":\"5671\"}],\"8883/tcp\":[{\"HostPort\":\"8883\"}]}}}"
                        },
                        "type": "docker",
                        "env": {
                            "storageFolder": {
                                "value": "/iotedge/storage/"
                            },
                            "OptimizeForPerformance": {
                                "value": "true"
                            },
                            "UsePersistentStorage": {
                                "value": "false"
                            }
                        },
                        "status": "running",
                        "restartPolicy": "always"
                    }
```
                
[result.txt](https://github.com/Azure/iotedge/files/4575671/result.txt)

**Comment by damonbarry** [comment]
@peirens-bart Which language SDK are you using in your A/B/C modules to send and receive messages? Also, what kind of hardware are you using to run IoT Edge? @varunpuranik

**Comment by peirens-bart** [comment]
I'm using c# with sdk Microsoft.Azure.Devices.Client 1.25.0
my host is a NUC 10 I7-107110U met 32GB Mem.
The hub logs are attachted.
What i noticed is that the hub is first receiving messages from module A and then start sending them in batch. why is the hub sending in Batch and not just forwarding the messages directly.

[output.txt](https://github.com/Azure/iotedge/files/4585061/output.txt)

**Comment by damonbarry** [comment]
@peirens-bart Thanks.
@varunpuranik Thoughts?

**Comment by peirens-bart** [comment]
@damonbarry , @varunpuranik did you had time to think about the issue?

**Comment by varunpuranik** [comment]
@peirens-bart -
I think what is happening in your case is that you have a fast sender A, and the receiver B is relatively slow. This is because the ingestion in EdgeHub is faster then dispatch to modules. This is a known issue right now. Because of this, the time diff between A to B keeps increasing, since one is processing at a faster rate than the other. 
Because B receives messages relatively slower, it sends messages at the same rate, and then C is able to keep up with that speed. So B to C duration seems stable. 

> What i noticed is that the hub is first receiving messages from module A and then start sending them in batch. why is the hub sending in Batch and not just forwarding the messages directly.
When EdgeHub receives messages, it stores them locally before dispatching them to the destination. This is currently by design, to guarantee that messages that are received are persisted and not lost. The batch you are seeing is for internal processing only, the messages are finally sent to the module one by one.

Since the behavior you are seeing is by design right now, I will close the issue. If there is anything else that you feel is a issue with the product, feel free to re-open the issue, or create a new one.

**Comment by peirens-bart** [comment]
@varunpuranik I'm sorry but I don't understand your explanation.

> it stores them locally before dispatching them to the destination

Why would the edgeHub doing this if using this setting, I understand that this will take time, but in my case, this is not needed because the events are useless if not immediately processed, so why storing them.

```
"UsePersistentStorage": {
    "value": "false"
}
```

You also said that this is because my module B is slow. how do you make this assumption?

> fast sender A, and the receiver B is relatively slow

2020-05-05 12:30:10.2712 | I | 80 | HertzReceiveActor.cs | 214 | Sending message with Id 6663414471546904576 and Type RfidReadEvent from Module tmreader with arguments [{"ANTENNAPORTNUMBER": "8"},{"RSSI": "-47"},{"TAG": "

2020-05-05 12:30:10.287 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received 1 message(s) from iwd-ubuntu-0001/tmreader

so sending from module A -> edgeHub = 16ms

2020-05-05 12:30:10.3974 | I | 44 | HertzReceiveActor.cs | 137 | Receive message with Id 6663414471546904576 and Type DequeueEvent from Module tmreader in Module flightcache with arguments [{"ANTENNAPORTNUMBER": "8"},{"RSSI": "-47"},{"TAG": "E2004102691201851680680F"},{"TAGLASTSEEN": "2020-05-05T12:30:10.2710555Z"},{"PHASE": "90"},{"FREQUENCY": "917500"},{"READERTEMPERATURE": "27"},{"READERNAME": "tmreader"},{"READERTIME": "19"},{"READERREADCOUNT": "1"},{"READERPERFORMANCE": "53"},{"ID": "6663414471546904576"}]

receive edgeHub -> receive module B = 287 ms

2020-05-05 12:30:10.3974 | I | 44 | HertzReceiveActor.cs | 214 | Sending message with Id 6663414471546904576 and Type DequeueEvent from Module flightcache with arguments [{"ANTENNAPORTNUMBER": "8"},{"RSSI": "-47"},{"TAG": "E2004102691201851680680F"},{"TAGLASTSEEN": "2020-05-05T12:30:10.2710555Z"},{"PHASE": "90"},{"FREQUENCY": "917500"},{"READERTEMPERATURE": "27"},{"READERNAME": "tmreader"},{"READERTIME": "19"},{"READERREADCOUNT": "1"},{"READERPERFORMANCE": "53"},{"ID": "6663414471546904576"},{"AMOUNTOFNEXTVALUES": "-1"}]

receive module B -> send to edgeHub = 0ms

2020-05-05 12:30:10.397 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received 1 message(s) from iwd-ubuntu-0001/flightcache

send module B -> receive edgeHub = 0ms

for what I can see the edgeHub is the bottleneck, in this case, is the edgeHUb with 303ms just for sending 1 message from module A to module B

as you can see in the attached log files the processing time of module B is almost always 0ms.

As for now, I made 1 module with my own "hub" based on Akka.net and every module is an actor, and now I have the correct performance without changing anything else to the code. I'm just left with a complex installer.
